services/monitor/src/ChainMonitor.ts

```diff
@@ -0,0 +1,281 @@
+    this.blockIntervalFactor = chainConfig.blockIntervalFactor;
+    assert(this.blockIntervalFactor > 1);
+    this.blockIntervalUpperLimit = chainConfig.blockIntervalUpperLimit;
+    this.blockIntervalLowerLimit = chainConfig.blockIntervalLowerLimit;
+    this.bytecodeNumberOfTries = chainConfig.bytecodeNumberOfTries;
+  }
+
+  start = async (): Promise<void> => {
+    try {
+      const lastBlockNumber = await this.sourcifyChain.getBlockNumber();
+      const startBlock =
+        this.startBlock !== undefined ? this.startBlock : lastBlockNumber;
+
+      this.chainLogger.info(`Starting monitor on block ${startBlock}`);
+
+      // Start polling
+      this.pollBlocks(startBlock);
+
+      // Listen to new blocks
+      this.on(NEW_BLOCK_EVENT, this.processBlockListener);
+    } catch (err: any) {
+      this.chainLogger.error(`Error starting monitor: ${err.message}`);
+    }
+  };
+
+  /**
+   * Stops the monitor after executing all pending requests.
+   */
+  stop = (): void => {
+    this.chainLogger.info(`Stopping monitor`);
+    this.off(NEW_BLOCK_EVENT, this.processBlockListener);
+  };
+
+  // Tries to get the next block by polling in variable intervals.
+  // If the block is fetched, emits a creation event, decreases the pause between blocks, and goes to next block
+  // If the block isn't there yet, it increases the pause between blocks.
+  private pollBlocks = async (startBlockNumber: number) => {
+    let currentBlockNumber = startBlockNumber;
+    const pollNextBlock = async () => {
+      try {
+        this.chainLogger.debug(
+          `Polling for block ${currentBlockNumber} with pause ${this.blockInterval}...`
+        );
+        const block = await this.sourcifyChain.getBlock(
+          currentBlockNumber,
+          true
+        );
+
+        if (block) {
+          this.adaptBlockPause("decrease");
+          currentBlockNumber++;
+          this.emit(NEW_BLOCK_EVENT, block);
+        } else {
+          this.adaptBlockPause("increase");
+        }
+
+        // Continue polling
+        setTimeout(pollNextBlock, this.blockInterval);
+      } catch (error: any) {
+        this.chainLogger.error(
+          `Error fetching block ${currentBlockNumber}: ${error?.message}`
+        );
+        this.adaptBlockPause("increase");
+        // Continue polling
+        setTimeout(pollNextBlock, this.blockInterval);
+      }
+    };
+
+    pollNextBlock();
+  };
+
+  // ListenerFunction
+  private processBlockListener = async (block: Block) => {
+    this.chainLogger.info(`Found block ${block.number}. Now processing`);
+
+    for (const tx of block.prefetchedTransactions) {
+      // TODO: Check factory contracts with traces
+      if (createsContract(tx)) {
+        const address = getCreateAddress(tx);
+        this.chainLogger.info(
+          `Found new contract in block ${block.number} with address ${address} created by tx ${tx.hash}`
+        );
+        this.processNewContract(tx.hash, address);
```

Comment:
Here I see that we process all contracts concurrently. Could we risk to overload our server? Maybe awaiting would be enought until we implement something to handle many compilations?  Was it without the await also before?

Jonestly I prefer it like this (without the await), maybe we can try leaving it like this and see if it causes many problems.